### ISSUE TITLE
Admin and Code Review

### ISSUE THREAD

**Issue by mguy-usgs** [usgs/geomag-imp#3]
here is feedback on admin and code review of https://github.com/usgs/geomag-imp

* recommend create a release candidate branch and/or tag as described at, this way you have a tagged version of everything that is reviewed and easy to make a tar ball of, all of which can be merged back into master.
https://github.com/usgs/best-practices/blob/master/software/reviews.md
don't worry about making the review issue, as I will make one and put all my comments into one, which you can split out to which parts you will take action on and which parts you will keep in back log as "recommendations"
*Licence.md - looks like there are dependencies on numpy, scipy, and other *py's, matplotlib, json-tricks, the licence(s) for these should be mentioned in your license.md file (e.g. for all python stuff maybe just say open source as defined https://www.python.org/psf/ , another example https://matplotlib.org/users/license.html, json-tricks is BSD http://json-tricks.readthedocs.io/en/latest/)
here is an example at least showing licenses for contributed data, a similar layout would exist for 3rd party code used https://github.com/usgs/pager/blob/master/LICENSE.md
or alternatively you could add something like this note saying see licenses in dependent libraries...
https://github.com/usgs/hazdev-leaflet/blob/master/LICENSE.md
*readme.md looks good, helpful details
* bin/make-imp-gp.py - i like the prerequisites and the what it does and does NOT do; I like the comments inline indicating units; there is little to no error handling, I strongly recommend adding in try/except blocks for at least the most generic exception but maybe a few others if they are known possibilities.
* there seems to be significant chunks of code (e.g. setup/config lines, creating arrays, etc.) duplicated between make-imp-gp.py and make-imp-sec.py consider making shared or utility code, so as to make changes/fixes in one place rather than multiple...
* in make_svsqdist.py just noting there is reference to cwbpub.cr.usgs.gov and port but this is all public so all good; again not try/except blocks, when working with things like streams and arrays/lists seems like a good idea
* data directory maybe nice to have the noaa format spec url in the files, but since it is in the code that produces them OK, or maybe add it to readme..this is minor https://www.ngdc.noaa.gov/IAGA/vdat/IAGA2002/iaga2002format.html

**Comment by mguy-usgs** [comment]
A few more comments with examples, to meet current standards and consistency:
* license.md correct link to https://www.usgs.gov/information-policies-and-instructions/copyrights-and-credits, 
* also move the disclaimer out of license.md to Disclaimer.md (example https://github.com/usgs/shakemap/blob/master/DISCLAIMER.md), and I also recommend adding the "approved" verbiage along with your tag reference into disclaimer.md, so basically you have both provisional (for any on going changes that may occur) as well as approved for this release.
Approved software verbiage:
"This software has been approved for release by the U.S. Geological Survey (USGS). Although the software has been subjected to rigorous review, the USGS reserves the right to update the software as needed pursuant to further analysis and review. No warranty, expressed or implied, is made by the USGS or the U.S. Government as to the functionality of the software and related material nor shall the fact of release constitute any such warranty. Furthermore, the software is released on condition that neither the USGS nor the U.S. Government shall be held liable for any damages resulting from its authorized or unauthorized use."

*add metadata file for the project, see example at: https://github.com/usgs/best-practices/blob/master/code.json

* make a release candidate tag for this, since you already did a pull request, you can add the tag on master after that pull request is merged... example might be : git tag -am 'vMAJOR.MINOR.PATCH-rc1' vMAJOR.MINOR.PATCH-rc1

**Comment by erigler-usgs** [comment]
Response to Administrative Software Review:
---------------------------------------------

Below I address each of @mguy-usgs's comments in the order they were provided:

1.  A [release candidate branch][1] was created, then a "Work in Progress" (WIP) pull request (PR) to the main GitHub repository was created. This is is actually where the Subject Matter Review was addressed. In the future, all reviews will be requested against a PR. Perhaps rather than prefixing the PR name with "WIP:", the prefix "RC:" will be used to help distinguish it from smaller WIPs. If you have an alternative recommendation, I would be very happy to consider it.
2.  Following the example of [Shakemap][2], we opted for a simpler LICENSE file that provided a link to a generic Python license statement, and encouraged the reader to look at dependencies for their own respective licenses. The major dependencies are listed in the README file.
3.  Thanks for the supportive words re. README file.
4.  The scripts make_*.py in the bin/ folder are not intended as user applications, but rather as demonstrations. This is made clear in the README file. Originally, I wanted little more than Bash script wrappers to [geomag-algorithms][3] (which already does a reasonable job of error handling), but there was a bug in its IAGA2002 factory at the time that was not easy to track down. Therefore, I wrote my own, probably less robust, but perfectly functional, pseudo-factory. Moving forward, I may just remove these scripts entirely, in favor of using notebooks for any and all demonstrations. For now, however, they serve their purpose, and have been proven to serve the purpose of at least one major customer, the NOAA Space Weather Prediction Center, who have been using this package into their own operations for over a year.
5.  I agree, there is substantial duplication of code between the different make_*.py scripts. I will refer back to my prior response, and note that these were never intended to be user applications, but just to demonstrate how one might use the library. Moving forward, if we keep these scripts, some effort will go into ensuring that the duplicate code is consolidated into some kind of "utility" package, which should lessen the likelihood that inconsistencies might be introduced to the logic of these interdependent scripts.
6.  I guess I am curious why you believe try-except blocks are especially useful for dealing with streams and arrays/lists. I am not against them at all, especially for conditional flow control (i.e., sometimes they are just much more efficient and/or comprehensible than complicated if-then-else constructs). I just don't think I've ever seen a reason to favor them just because I'm dealing with mildly complicated data structures.
7.  I link to the IAGA2002 file data definition in the README file. However, I also made [changes to the repository] [1] that removed and stopped tracking these data files. The user may well run the data/run_me.sh script and generate output files, but they will not be available via GitHub once this PR is merged.
8.  The new USGS copyright link is now used in the LICENSE file.
9.  I believe I have created a DISCLAIMER file that follows your recommendation.
10. I added a code.json file to the project.
11. I was planning to create a release tag as soon as this response is approved and I merge the PR with master. I know we already discussed this, but I guess I am still a little bit unsure about what you are requesting in your last comment when you suggest a tag that ends with '-rc1'. Or, are you saying that **next time** I should do this prior to submitting a "release candidate" for review?

Thank you so much for your careful and constructive review. I expect this will all go much more smoothly and quickly the next time, which may well be version 2.0 of this very project.


[1]: https://github.com/erigler-usgs/geomag-imp/tree/reconcile_peer_reviews
[2]: https://github.com/usgs/shakemap
[3]: https://github.com/usgs/geomag-algorithms

**Comment by mguy-usgs** [comment]
Good reconciliation. I found one question, more about thoughts than any action:
"why you believe try-except blocks are especially useful for dealing with streams and arrays/lists" mostly because you are more likely to get "array out of bound" sorts of exceptions

**Comment by erigler-usgs** [comment]
I am going to close this issue. I'm not sure why I did not do this almost two years ago after a favorable response from @mguy-usgs. I will also soon be addressing issues #4, #5, #6, #7, #8, and #9 submitted by @mhearne-usgs as a follow-up review in my attempt to finally get this project beyond "provisional" status.
